### PR TITLE
install-xattr: avoid accessing empty storage

### PR DESCRIPTION
**Pull request by thesamesam**
UBSAN reports:
```
install-xattr.c:124:16: runtime error: load of address 0x55555556d440 with insufficient space for an object of type 'char'
0x55555556d440: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  61 00 00 00
              ^
    #0 0x555555557a27 in copyxattr /home/sam/git/elfix/misc/install-xattr/install-xattr.c:124
    #1 0x555555556a4d in main /home/sam/git/elfix/misc/install-xattr/install-xattr.c:410
    #2 0x7ffff77c864f  (/usr/lib64/libc.so.6+0x2364f)
    #3 0x7ffff77c8708 in __libc_start_main (/usr/lib64/libc.so.6+0x23708)
    #4 0x555555557114 in _start (/home/sam/git/elfix/misc/install-xattr/install-xattr+0x3114)
```

Triggered with:
```
mkdir /tmp/a
touch /tmp/foo
./install-xattr -c /tmp/foo /tmp/foo2 /tmp/a
```

I don't see this with Clang or < GCC 12, but I do with GCC 13 (13.0.0_pre20230101 p5); I suspect it's because of object-size improvements.

Signed-off-by: Sam James <sam@gentoo.org>

### CODE REVIEW

**Comment by thesamesam** [comment]
@blueness

**Review by SoapGentoo — APPROVED** [review]

File: misc/install-xattr/install-xattr.c [hunk 1]
```diff
@@ -119,6 +119,11 @@ copyxattr(const char *source, const char *target)
 	lxattr = xmalloc(lsize);
 	xlistxattr(source, lxattr, lsize);
 
+	/* There's no xattrs at all. */
+	if (lsize == 0) {
+		return;
+	}
```
Comment (mgorny) [inline misc/install-xattr/install-xattr.c]: Brackets don't seem consistent with the coding style.

File: misc/install-xattr/install-xattr.c [hunk 2]
```diff
@@ -383,6 +386,7 @@ main(int argc, char* argv[])
 			if (!opts_target_directory) {
 				target = argv_copy[last];
 				if (stat(target, &s) != 0) {
+					free(mypath);
```
Comment (mgorny) [inline misc/install-xattr/install-xattr.c]: Why not free it immediately after `default:`? it doesn't seem to be used elsewhere.

In fact, given it's only used in one place, perhaps the whole variable can be scoped in `case 0:`. Also, the program is completely missing error handling for it.

Comment (thesamesam) [reply misc/install-xattr/install-xattr.c]: Oh, good spot.

Comment (mgorny) [reply misc/install-xattr/install-xattr.c]: You seem to have implemented my first paragraph but not the second. Any reason not to?

Comment (thesamesam) [reply misc/install-xattr/install-xattr.c]: Just no time yet - that's why the comment isn't resolved. I think your suggestion's a good one.

File: misc/install-xattr/install-xattr.c [hunk 3]
```diff
@@ -352,7 +352,7 @@ main(int argc, char* argv[])
 		case -1:
 			err(1, "fork() failed");
 
-		case 0:
+		case 0: ;
 			char *mypath = realpath("/proc/self/exe", NULL); /* path to argv[0] */
```
Comment (SoapGentoo) [inline misc/install-xattr/install-xattr.c:356]: can't declare a variable in a case statement without another scope?

Comment (thesamesam) [reply misc/install-xattr/install-xattr.c:356]: Clang hates it:
```
$ make CC=clang-16 CFLAGS="-std=gnu17 -O2 -pedantic -fsanitize=address -ggdb3" LDFLAGS="-fsanitize=address -ggdb3"
clang-16 -std=gnu17 -O2 -pedantic -fsanitize=address -ggdb3  -fsanitize=address -ggdb3 -o install-xattr install-xattr.c
install-xattr.c:356:4: error: expected expression
                        char *mypath = realpath("/proc/self/exe", NULL); /* path to argv[0] */
                        ^
install-xattr.c:361:20: error: use of undeclared identifier 'mypath'; did you mean 'path'?
                        install = which(mypath, portage_helper_canpath);
                                        ^~~~~~
                                        path
install-xattr.c:249:8: note: 'path' declared here
        char *path;                    /* path to the target file                                      */
              ^
install-xattr.c:362:9: error: use of undeclared identifier 'mypath'; did you mean 'path'?
                        free(mypath);
                             ^~~~~~
                             path
install-xattr.c:249:8: note: 'path' declared here
        char *path;                    /* path to the target file                                      */
              ^
3 errors generated.
make: *** [Makefile:8: install-xattr] Error 1
```

```
$ make CC=gcc CFLAGS="-std=gnu17 -O2 -pedantic -fsanitize=address -ggdb3" LDFLAGS="-fsanitize=address -ggdb3"
gcc -std=gnu17 -O2 -pedantic -fsanitize=address -ggdb3  -fsanitize=address -ggdb3 -o install-xattr install-xattr.c
install-xattr.c: In function ‘main’:
install-xattr.c:356:25: warning: a label can only be part of a statement and a declaration is not a statement [-Wpedantic]
  356 |                         char *mypath = realpath("/proc/self/exe", NULL); /* path to argv[0] */
      |                         ^~~~
```

Comment (SoapGentoo) [reply misc/install-xattr/install-xattr.c:356]: wrap it in a `{}` scope

File: misc/fix-gnustack/tests/Makefile.am [hunk 4]
```diff
@@ -5,11 +5,11 @@ noinst_PROGRAMS = bad-gnustack
 EXTRA_DIST = gnustacktest.sh
 
 bad-gnustack.s: bad-gnustack.c
-	$(CC) -S $<
+	$(CC) $(CFLAGS) $(LDFLAGS) -S $<
```
Comment (SoapGentoo) [inline misc/fix-gnustack/tests/Makefile.am]: CPPFLAGS?

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: misc/fix-gnustack/tests/Makefile.am (hunk 4 above)
Comment (SoapGentoo) [reply misc/fix-gnustack/tests/Makefile.am]: also, not sure the LDFLAGS makes a lot of sense

**Review by SoapGentoo — APPROVED** [review]

**Pull request closed without merging**